src/qpg7015m/spinel_extension.cpp

```diff
@@ -0,0 +1,221 @@
+
+// #include "utils/uart.h"
+#include "ncp_base.hpp"
+#include <openthread/thread_ftd.h>
+#include "lib/spinel/spinel.h"
+
+extern "C" void    qorvoNeighbourTableChanged(otNeighborTableEvent aEvent, const otNeighborTableEntryInfo *aEntryInfo);
+extern "C" otError otPlatUartSend(const uint8_t *aBuf, uint16_t aBufLength);
+
+#define SPINEL_PROP_VENDOR_CHILD_TABLE SPINEL_PROP_VENDOR__BEGIN + 0
+#define SPINEL_PROP_VENDOR_NEIGHBOR_TABLE_CALLBACK SPINEL_PROP_VENDOR__BEGIN + 1
+
+#include "ncp_hdlc.hpp"
+#include "common/new.hpp"
+
+class NcpVendorUart : public ot::Ncp::NcpHdlc
+{
+    static int SendHdlc(const uint8_t *aBuf, uint16_t aBufLength)
+    {
+        otPlatUartSend(aBuf, aBufLength);
+        return 0;
+    }
+
+public:
+    NcpVendorUart(ot::Instance *aInstance)
+        : ot::Ncp::NcpHdlc(aInstance, &NcpVendorUart::SendHdlc)
+        , mRegisterCallback(false)
+    {
+    }
+
+    void RegisterNeighborTableCallback(bool enable);
+    bool mRegisterCallback;
+
+private:
+    // Add public/private methods or member variables
+};
+
+namespace ot {
+namespace Ncp {
+
+otError NcpBase::VendorCommandHandler(uint8_t aHeader, unsigned int aCommand)
+{
+    otError error = OT_ERROR_NONE;
+
+    switch (aCommand)
+    {
+        // TODO: Implement your command handlers here.
+
+    default:
+        error = PrepareLastStatusResponse(aHeader, SPINEL_STATUS_INVALID_COMMAND);
+    }
+
+    return error;
+}
+
+void NcpBase::VendorHandleFrameRemovedFromNcpBuffer(Spinel::Buffer::FrameTag aFrameTag)
+{
+    // This method is a callback which mirrors `NcpBase::HandleFrameRemovedFromNcpBuffer()`.
+    // It is called when a spinel frame is sent and removed from NCP buffer.
+    //
+    // (a) This can be used to track and verify that a vendor spinel frame response is
+    //     delivered to the host (tracking the frame using its tag).
+    //
+    // (b) It indicates that NCP buffer space is now available (since a spinel frame is
+    //     removed). This can be used to implement reliability mechanisms to re-send
+    //     a failed spinel command response (or an async spinel frame) transmission
+    //     (failed earlier due to NCP buffer being full).
+
+    OT_UNUSED_VARIABLE(aFrameTag);
+}
+
+otError NcpBase::VendorGetPropertyHandler(spinel_prop_key_t aPropKey)
+{
+    otError error = OT_ERROR_NONE;
+
+    switch (aPropKey)
+    {
+        // TODO: Implement your property get handlers here.
+        //
+        // Get handler should retrieve the property value and then encode and write the
+        // value into the NCP buffer. If the "get" operation itself fails, handler should
+        // write a `LAST_STATUS` with the error status into the NCP buffer. `OT_ERROR_NO_BUFS`
+        // should be returned if NCP buffer is full and response cannot be written.
+    case SPINEL_PROP_VENDOR_NEIGHBOR_TABLE_CALLBACK:
+    {
+        bool enable = ((NcpVendorUart *)(ot::Ncp::NcpBase::GetNcpInstance()))->mRegisterCallback;
```

Comment:
I think can do this instead:
```suggestion
        bool enable = ((NcpVendorUart *)(this)->mRegisterCallback;
```